Feature sdlp hub spot sync

matthew-ewing opened the pull request:
Draft for LPC HS jobs

shaneLPDev committed:
- Added thecrossingchurch.CustomJobs files

shaneLPDev committed:
Adding files from Private Repo

shaneLPDev committed:
- Adjusting

shaneLPDev committed:
- Readjusting

shaneLPDev committed:
- adjustments

matthewelp committed:
- A bunch of updates

shaneLPDev committed:
- Added ; where needed

matthewelp committed:
- More changes

matthewelp committed:
- Fixes

matthewelp committed:
- More fixes

shaneLPDev committed:
- Added using System.Text;

shaneLPDev committed:
- Corrected errors

matthewelp committed:
- Maybe fix the queries

Jon-Corey commented on Rock.code-workspace:
```diff
@@ -72,7 +72,7 @@
 		"html.format.wrapLineLength": 0,
 		"editor.codeActionsOnSave": {
-			"source.fixAll.eslint": true
+			"source.fixAll.eslint": "explicit"
```
What does this change do? Do we need it? Does it make sense for it to be a part of the HubSpot changes?

Jon-Corey commented on org.lakepointe.Reporting/org.lakepointe.Reporting.csproj:
```diff
@@ -55,8 +55,8 @@
     <Compile Include="Reporting\DataFilter\Group\TagFilter.cs" />
-    <Compile Include="Reporting\DataFilter\Person\FirstAttendanceOnCampusFilter.cs" />
     <Compile Include="Reporting\DataFilter\Person\PersonInGroupAttribute.cs" />
+    <Compile Include="Reporting\DataFilter\Person\FirstAttendanceOnCampusFilter.cs" />
```
Why was this change made? Does this need to be a part of the HubSpot changes?

Jon-Corey commented on org.thecrossingchurch.CustomJobs/Jobs/HubspotIntegrationMatching.cs:
```diff
@@ -0,0 +1,688 @@
+        {
+            JobDataMap dataMap = context.JobDetail.JobDataMap;
+
+            // Get HS API Token
```
Can we change "HS" to "HubSpot" here? That'll make it clearer to our future selves what API Token this is without having to read through the code.
My preference is that comments and variable names should rarely contain Acronyms or shortened words. The extra characters don't cost anything and they serve to add clarity. The exception would be acronyms that most developers would know, like API or HTML.

Jon-Corey commented on org.thecrossingchurch.CustomJobs/Jobs/HubspotIntegrationPatching.cs:
```diff
@@ -88,23 +88,12 @@ public virtual void Execute(IJobExecutionContext context)
             JobDataMap dataMap = context.JobDetail.JobDataMap;
 
-            //Bearer Token, but I didn't change the Attribute Key
+            // Get HS API Token
```
Can we change "HS" to "HubSpot" here too?

Jon-Corey commented on org.thecrossingchurch.CustomJobs/Jobs/HubspotIntegrationPatching.cs:
```diff
@@ -0,0 +1,389 @@
+        {
+            JobDataMap dataMap = context.JobDetail.JobDataMap;
+
+            // Get HS API Token
```
Can we change "HS" to "HubSpot" here too?

shaneLPDev committed:
- Added a record count to Debug Line

shaneLPDev committed:
- Fix to Null Value on Date Attributes

shaneLPDev committed:
- Fixed Errors And Added Status Messaging

shaneLPDev committed:
- Bug fixes

- Nullable object must have a value exception fixed.
- Number of contacts to patch fixed.

shaneLPDev committed:
- Added fix to not attempt to add duplicate email address to HS

shaneLPDev committed:
- Added feature Skip Email If Not Proper Format

Will not attempt to add email address to HubSpot if email address does not contain "@" and does not use one of 536 TLDs.

shaneLPDev committed:
- Bug fixes

Jon-Corey commented on org.thecrossingchurch.CustomJobs/Jobs/HubspotIntegrationMatching.cs:
```diff
@@ -0,0 +1,718 @@
+        }
+
+        /// <summary>
+        /// Job that will run quick SQL queries on a schedule.
```
This XML code comment doesn't seem to be accurate.

Jon-Corey commented on org.thecrossingchurch.CustomJobs/Jobs/HubspotIntegrationMatching.cs:
```diff
@@ -0,0 +1,718 @@
+        /// 
+        /// Called by the <see cref="IScheduler" /> when a
+        /// <see cref="ITrigger" /> fires that is associated with
+        /// the <see cref="IJob" />.
```
This part of the XML code comment seems overly complicated and not really applicable to this job specifically. This seems like something that'd be a better fit on the interface itself. Can we just remove this section?

Jon-Corey commented on org.thecrossingchurch.CustomJobs/Jobs/HubspotIntegrationMatching.cs:
```diff
@@ -0,0 +1,718 @@
+        {
+            Debug.WriteLine( "***********************************************************************************************************************************************" );
+            Debug.WriteLine( "********************************************* HubSpot Integration Matching Job Start **********************************************************" );
+            Debug.WriteLine( "*************************************************" + DateTime.Now.ToString() + "**************************************************************" );
```
Debug statements should either be removed or commented out.

Jon-Corey commented on org.thecrossingchurch.CustomJobs/Jobs/HubspotIntegrationMatching.cs:
Why are we validating emails like this? This way of doing things leaves a few gaps. C# has a built-in way to validate emails that would be simpler and more consistent: [link](https://stackoverflow.com/questions/1365407/c-sharp-code-to-validate-email-address).

Jon-Corey commented on org.thecrossingchurch.CustomJobs/Jobs/HubspotIntegrationMatching.cs:
This doesn't seem to ever be incremented, so it is always 0. This defeats the purpose of having this number. Can we either add code to increment it or remove it entirely?

Jon-Corey commented on org.thecrossingchurch.CustomJobs/Jobs/HubspotIntegrationMatching.cs:
```diff
@@ -0,0 +1,718 @@
+        /// <see cref="ITrigger" /> fires that is associated with
+        /// the <see cref="IJob" />.
+        /// </summary>
+        public virtual void Execute(IJobExecutionContext context)
```
Rock's [style guide](https://github.com/SparkDevNetwork/Rock/wiki/Contributing-to-Rock#styleguides) dictates spaces within parenthesis when they contain a value. This should be `public virtual void Execute( IJobExecutionContext context )`.

There are a few other places where this also needs to be updated.

Jon-Corey commented on org.thecrossingchurch.CustomJobs/Jobs/HubspotIntegrationMatching.cs:
Can we change the `_context` to use a using statement so that it is automatically disposed of? ([video](https://youtu.be/C4qzrFeYHgs?feature=shared))

Jon-Corey commented on org.thecrossingchurch.CustomJobs/Jobs/HubspotIntegrationMatching.cs:
Can we change the `personService` and `personSearchKeyService` to use using statements so that they are automatically disposed of?

Jon-Corey commented on org.thecrossingchurch.CustomJobs/Jobs/HubspotIntegrationMatching.cs:
Can we add error handling to this so that if HubSpot is unresponsive, our key is no longer valid, or some other issue happens with this request the job fails gracefully instead of trying to use bad/nonexistent data?
This could be in the form of checking the status code of the response to make sure it starts with a 2 and throwing an error if it doesn't.

Jon-Corey commented on org.thecrossingchurch.CustomJobs/Jobs/HubspotIntegrationMatching.cs:
The `String.IsNullOrEmpty` here works, but it is general practice to use the `string` alias. So `String.IsNullOrEmpty` should be `string.IsNullOrEmpty`.

Also, this is a preference thing, but I am not a fan of using `!` to flip conditionals. I would rather we use `== false` instead because it is more clear and easier to read.

Jon-Corey commented on org.thecrossingchurch.CustomJobs/Jobs/HubspotIntegrationMatching.cs:
If `key` is blank, the job can't function properly. Can we add some validation here that throws an exception and stops the job from running if `key` is blank?

Jon-Corey commented on org.thecrossingchurch.CustomJobs/Jobs/HubspotIntegrationMatching.cs:
```diff
@@ -0,0 +1,718 @@
+    [TextField("AttributeKey", "The attribute key for the global attribute that contains the HubSpot API Key. The attribute must be encrypted.", true, "HubspotAPIKeyGlobal")]
+    [TextField("Business Unit", "Hubspot Business Unit value", true, "0")]
+    [TextField("Potential Matches File Name", "Name of the file for this job to list potential matches for cleaning", true, "Potential_Matches")]
+    public class HubspotIntegrationMatching : IJob
```
HubSpot is spelled with Pascal case. This file has inconsistent capitalization in a few places. Can they all be updated to "HubSpot"?

Jon-Corey replied:
This job needs to inherit from `RockJob`. `IJob` will no longer be supported in v16.

Jon-Corey commented on org.thecrossingchurch.CustomJobs/Jobs/HubspotIntegrationMatching.cs:
Rock should be capitalized here since it is a proper noun.

Jon-Corey commented on org.thecrossingchurch.CustomJobs/Jobs/HubspotIntegrationMatching.cs:
It looks like `i` is just used for getting the nth contact. Could we update this to a `foreach` loop for clarity?

Jon-Corey commented on org.thecrossingchurch.CustomJobs/Jobs/HubspotIntegrationMatching.cs:
Why are we using a custom `HubspotPropertyUpdate` class when a dictionary could fulfill the same purpose? Can we swap this out for a dictionary?

Jon-Corey commented on org.thecrossingchurch.CustomJobs/Jobs/HubspotIntegrationMatching.cs:
This method doesn't use `_context`. Can we just remove it?

Jon-Corey commented on org.thecrossingchurch.CustomJobs/Jobs/HubspotIntegrationMatching.cs:
This should probably be an `else if` since there is never going to be a time where both `if` statements fire.

Jon-Corey commented on org.thecrossingchurch.CustomJobs/Jobs/HubspotIntegrationMatching.cs:
I've mentioned this in a previous comment, but I'd prefer `hasMultiEmail == false` over `!hasMultiEmail`. Spelling out the `== false` is more clear, which would make it easier to read later.

Jon-Corey commented on org.thecrossingchurch.CustomJobs/Jobs/HubspotIntegrationMatching.cs:
It seems like it would be more efficient to search for a phone number using the phone number service instead of the person service. Have you looked into the performance implications of each option?

Jon-Corey replied:
Looked into the performance with Shane. Performance difference was small enough we can ignore this change.

Jon-Corey commented on org.thecrossingchurch.CustomJobs/Jobs/HubspotIntegrationMatching.cs:
Could we rename `CustomEquals` to something like `HasValueAndEquals` or something else that describes what it does. Methods should have names that describe what they do in order to make code more readable.

Jon-Corey commented on org.thecrossingchurch.CustomJobs/Jobs/HubspotIntegrationMatching.cs:
`.ToStringSafe()` does not work if the string is null. C# does not allow anything after the dot if the value is null. If `email` is null here, that would cause a `NullReferenceException`. And converting it to a string is not necessary since it already is a string.

This could be changed to:
```csharp
if ( string.IsNullOrEmpty( contact.properties.email ) == false )
{
    hsPersonEmails.Add( contact.properties.email.ToLower() );
}
```

Jon-Corey commented on org.thecrossingchurch.CustomJobs/Jobs/HubspotIntegrationMatching.cs:
This method doesn't use `_context`. Can we just remove it?

Jon-Corey commented on org.thecrossingchurch.CustomJobs/Jobs/HubspotIntegrationMatching.cs:
Can we add criteria for only getting records with a `RecordTypeValueId` of 1 (Person) since we don't need businesses, nameless people, Rest keys, etc.

Can we also filter out people where `IsDeceased` is 1 (true)? The record status should be inactive for all of these people, but that isn't a guarantee.

Also, can we filter out people where `IsLockedAsChild` is 1 (true) since they are not classified as adults in our database?

Jon-Corey commented on org.thecrossingchurch.CustomJobs/Jobs/HubspotIntegrationMatching.cs:
Here's another place that needs to be refactored to remove `.ToStringSafe()`.

---

And this is just a preference, but I'd prefer that all if statements have the contents on a separate line from the if statement itself.

So, the following would be incorrect:
```csharp
if ( condition == true ) { doSomething(); }
```

And this would be preferred:
```csharp
if ( condition == true )
{
    doSomething();
}
```

This is another thing where adding those additional characters improves readability. Code is meant to be read by humans, and adding additional characters is free. My preference is pretty much always going to be to add additional characters if it can improve readability.

Jon-Corey commented on org.thecrossingchurch.CustomJobs/Jobs/HubspotIntegrationMatching.cs:
This is redundant (hence the comment). Can we remove it?

Jon-Corey commented on org.thecrossingchurch.CustomJobs/Jobs/HubspotIntegrationMatching.cs:
Can we add a using statement for System.IO so that we don't have to spell it out each time? That'd help to improve readability.

Jon-Corey commented on org.thecrossingchurch.CustomJobs/Jobs/HubspotIntegrationMatching.cs:
The `CustomEquals` method already does a null check, so I believe the `num != null` is redundant and can be removed.

Jon-Corey commented on org.thecrossingchurch.CustomJobs/Jobs/HubspotIntegrationMatching.cs:
Can we change this Rock URL to use the Internal Application Root global attribute so that this still works if we run this in a dev environment or if we change our domain?

To get a value from the global attribute cache in C#:
```csharp
GlobalAttributesCache.Get().GetValue( "InternalApplicationRoot" );
```

Jon-Corey commented on org.thecrossingchurch.CustomJobs/Jobs/HubspotIntegrationMatching.cs:
The only thing using this class is the Patching job. Can we move it to the Patching job?

Jon-Corey commented on org.thecrossingchurch.CustomJobs/Jobs/HubspotIntegrationMatching.cs:
The only thing using this class is the Patching job. Can we move it to the Patching job?

Jon-Corey commented on org.thecrossingchurch.CustomJobs/Jobs/HubspotIntegrationMatching.cs:
This is used in both jobs. I'd prefer that it either be duplicated in both or that it be in a separate file that both jobs can access. Public classes should not live in jobs.

Jon-Corey commented on org.thecrossingchurch.CustomJobs/Jobs/HubspotIntegrationMatching.cs:
This class is only used in the Matching job. Can we mark it as private?

Jon-Corey commented on org.thecrossingchurch.CustomJobs/Jobs/HubspotIntegrationMatching.cs:
This is used in both jobs. I'd prefer that it either be duplicated in both or that it be in a separate file that both jobs can access. Public classes should not live in jobs.

Jon-Corey commented on org.thecrossingchurch.CustomJobs/Jobs/HubspotIntegrationPatching.cs:
```diff
@@ -0,0 +1,442 @@
+    )]
+    [BooleanField("Include TMBT", defaultValue: false)]
+    [AccountField("Financial Account", "If syncing a total amount given which fund should we sync from")]
+    public class HubspotIntegrationPatching : IJob
```
This also needs to inherit from `RockJob`.

Jon-Corey commented on org.thecrossingchurch.CustomJobs/Jobs/HubspotIntegrationPatching.cs:
```diff
@@ -0,0 +1,442 @@
+            //tmbtProps = propsQry.results.Where(p => p.groupName == "rock_tmbt_information").ToList(); // ---------- Commented 6/25 sdlp
+            //Business Unit hs_all_assigned_business_unit_ids
+            //Save a list of the ones that are Rock attributes
+            //var attrs = props.Where(p => p.label.Contains("Rock Attribute ")).ToList(); // commented sdlp 7/15/24
```
Please remove commented out code since it doesn't add to the readability of the code that is currently being used.

Jon-Corey commented on org.thecrossingchurch.CustomJobs/Jobs/HubspotIntegrationPatching.cs:
```diff
@@ -0,0 +1,442 @@
+            contacts = new List<HSContactResult>();
+            request_count = 0;
+
+            // TODO get all properties...
```
Is this To Do comment resolved? If so, can we remove it? If not, what is left to do and does it need to be done in the initial release?

Jon-Corey commented on org.thecrossingchurch.CustomJobs/Jobs/HubspotIntegrationPatching.cs:
```diff
@@ -0,0 +1,442 @@
+            PersonAliasService pa_svc = new PersonAliasService(_context);
+            FinancialTransactionService ft_svc = new FinancialTransactionService(_context);
+            AttributeValueService av_svc = new AttributeValueService(_context);
+            var dataViewService = new DataViewService( _context );
```
Can these services be converted to using using statements so that they are disposed of properly if there is an error?

Jon-Corey commented on org.thecrossingchurch.CustomJobs/Jobs/HubspotIntegrationPatching.cs:
```diff
@@ -0,0 +1,442 @@
+            FinancialTransactionService ft_svc = new FinancialTransactionService(_context);
+            AttributeValueService av_svc = new AttributeValueService(_context);
+            var dataViewService = new DataViewService( _context );
+            var enewsDV = dataViewService.Get( "e4f1db79-63c7-41ca-ab45-6ed6b16feb0e" ); // ENews DataView Id: 2882, Guid: e4f1db79-63c7-41ca-ab45-6ed6b16feb0e
```
Can we make this a job setting instead of hard-coding the Guid. It is generally not good practice to hard-code stuff that can change.

Jon-Corey commented on org.thecrossingchurch.CustomJobs/Jobs/HubspotIntegrationPatching.cs:
`.ToStringSafe()` will need to be replaced in a few places in this file too.

Jon-Corey commented on org.thecrossingchurch.CustomJobs/Jobs/HubspotIntegrationPatching.cs:
Is this the correct Id for production?

Jon-Corey commented on org.thecrossingchurch.CustomJobs/Jobs/HubspotIntegrationPatching.cs:
Doesn't matter much since this is a debug statement, but this counter will be off by one since `i` starts at 0.

These debug statements should probably either be removed or commented out.

Jon-Corey commented on org.thecrossingchurch.CustomJobs/Jobs/HubspotIntegrationMatching.cs:
```diff
@@ -0,0 +1,718 @@
+using System.Threading;
+using OfficeOpenXml.FormulaParsing.LexicalAnalysis;
+using System.Text;
+using DotLiquid.Tags;
```
It looks like the `OfficeOpenXml.FormulaParsing.LexicalAnalysis` and `DotLiquid.Tags` using statements aren't being used. Can we remove them?

Jon-Corey commented on org.thecrossingchurch.CustomJobs/Jobs/HubspotIntegrationPatching.cs:
```diff
@@ -0,0 +1,442 @@
+using Quartz.Impl.Matchers;
+using System.Text;
+using Rock.Lava.RockLiquid.Blocks;
+using RestSharp.Extensions;
```
It looks like there's a bunch of using statements here that aren't being used. Can we remove them?

Jon-Corey commented on org.thecrossingchurch.CustomJobs/Jobs/HubspotIntegrationMatching.cs:
```diff
@@ -0,0 +1,718 @@
+        private string key { get; set; }
+        private List<HSContactResult> contacts { get; set; }
+        private int request_count { get; set; }
+        private string businessUnit { get; set; }
```
The names of these properties violate naming rules. Please update them.
- `key` -> `Key`
- `contacts` -> `Contacts`
- `request_count` -> `RequestCount`

And `businessUnit` has 0 references. So can we remove it?

Jon-Corey commented on org.thecrossingchurch.CustomJobs/Jobs/HubspotIntegrationPatching.cs:
```diff
@@ -0,0 +1,442 @@
+        private string key { get; set; }
+        private List<HSContactResult> contacts { get; set; }
+        private int request_count { get; set; }
+        private string businessUnit { get; set; }
```
The names of these properties violate naming rules. Please update them.
- `key` -> `Key`
- `contacts` -> `Contacts`
- `request_count` -> `RequestCount`

And `businessUnit` has 0 references. So can we remove it?

Jon-Corey commented on org.thecrossingchurch.CustomJobs/Jobs/HubspotIntegrationPatching.cs:
```diff
@@ -0,0 +1,442 @@
+            props = propsQry.results;
+            WriteToLog( "Is propsQry Null? : " + propsQry.IsNull().ToString() );
+            WriteToLog( "Is propsQry.results Null? : " + propsQry.results.IsNull().ToString() );
+            WriteToLog( "Is props Null? : " + props.IsNull().ToString() );
```
The `WriteToLog` calls should probably be commented back out since they are used for debugging. These ones also need to be updated to replace the non-functional `.IsNull()` (which is also marked as obsolete).

Jon-Corey commented on org.thecrossingchurch.CustomJobs/Jobs/HubspotIntegrationPatching.cs:
Can we add null handling to this in case the defined value doesn't have these attribute values?

Jon-Corey commented on org.thecrossingchurch.CustomJobs/Jobs/HubspotIntegrationPatching.cs:
This is another place where we've got an abbreviation that is making the code harder to read. Please update the variable names to remove the abbreviations (e.g. `hsSyncDv` -> `hubSpotSyncDefinedValue`). The extra characters don't cost anything, and they improve readability.

Jon-Corey commented on org.thecrossingchurch.CustomJobs/Jobs/HubspotIntegrationPatching.cs:
The `date` variable here is not used anywhere. Can we just remove it?

Jon-Corey commented on org.thecrossingchurch.CustomJobs/Jobs/HubspotIntegrationPatching.cs:
This is a long Entity Framework query with a lot of distinct steps. Can you add some line breaks to improve readability?
And `.ToStringSafe()` will need to be replaced.

Jon-Corey commented on org.thecrossingchurch.CustomJobs/Jobs/HubspotIntegrationPatching.cs:
This code shouldn't require using a try/catch. If there are exceptions, they need to be fixed, not ignored. Can you remove the try/catch and fix any potential errors (looks like this might just need some null handling).

Jon-Corey commented on org.thecrossingchurch.CustomJobs/Jobs/HubspotIntegrationPatching.cs:
Please swap out the `!mobile.IsUnlisted` for `mobile.IsUnlisted == false` for readability.

Jon-Corey commented on org.thecrossingchurch.CustomJobs/Jobs/HubspotIntegrationPatching.cs:
This line is a bit overcomplicated and has some redundant parts. It could be simplified as:
```csharp
string email = person.Email ?? "";
```

The original code says "Set email variable to the person's email if it is not blank. If the person's email is blank, set the email variable to blank instead of the person's email". The new code says "Set the email variable to the person's email, but set it to blank if it is null".

Jon-Corey commented on org.thecrossingchurch.CustomJobs/Jobs/HubspotIntegrationPatching.cs:
Even though we're doing a null check in the previous code, it's typically good practice to check for null or empty instead of just empty. So this could be re-written as:
```csharp
if ( string.IsNullOrEmpty( email ) == false )
```

Jon-Corey commented on org.thecrossingchurch.CustomJobs/Jobs/HubspotIntegrationPatching.cs:
Email validation is built into C#. We should use the built-in validation instead of writing our own. See the same comment on the other job for details.

Jon-Corey commented on org.thecrossingchurch.CustomJobs/Jobs/HubspotIntegrationPatching.cs:
This query is not the standard way of doing a query with Entity Framework. Can it be updated to use the StepService to do this query instead?

Jon-Corey commented on org.thecrossingchurch.CustomJobs/Jobs/HubspotIntegrationPatching.cs:
This way of calling `ConvertDate` on `stepResult` is unnecessarily complex. Can you simplify it for the sake of readability?

Jon-Corey commented on org.thecrossingchurch.CustomJobs/Jobs/HubspotIntegrationPatching.cs:
The standard name for an Exception variable in a try/catch is `e`. Can you update this to follow the standard?

Jon-Corey commented on org.thecrossingchurch.CustomJobs/Jobs/HubspotIntegrationPatching.cs:
First, this seems like a very simple string. Why are we using `StringBuilder` here? We could use string interpolation (`$"Some text { someVariable }"`) and make this a lot more readable.

Second, why does it matter what the last person processed was? Are we going to use this info for anything? If not, can we remove it?

Jon-Corey commented on org.thecrossingchurch.CustomJobs/Jobs/HubspotIntegrationPatching.cs:
The `DateTime` object has a `Date` property which is the `DateTime`'s date at midnight. These two lines could be simplified as:
```csharp
var d = date.Value.Date.Subtract(new DateTime(1970, 1, 1)).TotalSeconds * 1000;
```

Jon-Corey commented on org.thecrossingchurch.CustomJobs/packages.config:
```diff
@@ -0,0 +1,8 @@
+  <package id="Newtonsoft.Json" version="13.0.1" targetFramework="net472" />
+  <package id="RestSharp" version="105.2.3" targetFramework="net472" requireReinstallation="true" />
+  <package id="SquaredUp.HubSpot.NET" version="0.6.17" targetFramework="net472" requireReinstallation="true" />
+  <package id="System.Diagnostics.DiagnosticSource" version="4.7.1" targetFramework="net472" requireReinstallation="true" />
```
The version of `System.Diagnostics.DiagnosticSource` that the HubSpot project is using (v4.7.1) is behind the version that the rest of Rock is using (v5.0.0). Having different versions of the same package can cause issues. Can we update this to v5.0.0?

Same thing for `EntityFramework`. Rock's core code is using v6.4.4, whereas the HubSpot project is using v6.1.3. Can we update this to be in line with core too?

Jon-Corey commented on org.thecrossingchurch.CustomJobs/Jobs/HubspotIntegrationPatching.cs:
The `.Select( item => item)` is redundant here. Not sure what this was trying to accomplish. Removing it will yield the same result.

Jon-Corey commented on org.thecrossingchurch.CustomJobs/Jobs/HubspotIntegrationPatching.cs:
`HashSet`s aren't meant for this use case. After a quick test, I saw that a `List` is over two times faster for this use case. Can we swap this out for a `List`?

matthew-ewing replied:
We used a HashSet here because our use case was to check if a person is in the eNews data view quickly. We use the .contains() method a few lines below..

https://github.com/lakepointe/Rock-LPC/pull/12/files/1db4f85efd0a2ea4da4343e5b734838c089a5174#diff-71af37808a48c098da3cf5525d40257e4fa34ec71afa197fed2799a54df6cb56R305

List contains should be order n time, whereas hashset contains should be order 1

matthew-ewing replied:
I can't speak for other instances where HashSet was used, but I specifically asked Shane to implement it here.

matthew-ewing replied:
https://david.gardiner.net.au/2021/10/fastest-contains.html

Jon-Corey replied:
I tested a very similar scenario on a .NET Framework console app and got a consistent performance increase while using a `List` or `Array` over a `HashSet`. The difference may come from adding items to the iterable. `HashSet` has some additional overhead when adding items, which we are doing a lot of. I'd be happy to show you my test scenario in our dev meeting and talk about this.

Jon-Corey replied:
Actually, I just realized the `Contains` calls were inside some pretty big loops and getting called a lot more than I thought. `HashSet` is the right choice here. I'll resolve this conversation and delete the others.

matthew-ewing replied:
<img src="https://media4.giphy.com/media/oe33xf3B50fsc/giphy.gif"/>

matthew-ewing replied:
Jeff said our conflict wasn't explosive, so I decided to add an explosion here.

Jon-Corey commented on org.thecrossingchurch.CustomJobs/Jobs/HubspotIntegrationPatching.cs:
`HashSet`s aren't meant for this use case. After a quick test, I saw that a `List` is over two times faster for this use case. Can we swap this out for a `List`?

matthew-ewing replied:
Appears this is used for .contains() also

Jon-Corey commented on org.thecrossingchurch.CustomJobs/Jobs/HubspotIntegrationMatching.cs:
`HashSet`s aren't meant for this use case. After a quick test, I saw that a `List` is over two times faster for this use case. Can we swap these out for `List`s?

matthew-ewing replied:
Appears these are used for .contains() also

Jon-Corey commented on org.thecrossingchurch.CustomJobs/Jobs/HubspotIntegrationPatching.cs:
```diff
@@ -0,0 +1,442 @@
+        private int request_count { get; set; }
+        private string businessUnit { get; set; }
+
+        /// <summary> 
```
There's some extra white space at the end of this line. Can you remove it?

Jon-Corey commented on org.thecrossingchurch.CustomJobs/Jobs/HubspotIntegrationPatching.cs:
```diff
@@ -0,0 +1,442 @@
+        /// 
+        /// Called by the <see cref="IScheduler" /> when a
+        /// <see cref="ITrigger" /> fires that is associated with
+        /// the <see cref="IJob" />.
```
This part of the XML code comment seems overly complicated and not really applicable to this job specifically. This seems like something that'd be a better fit on the interface itself. Can we just remove this section?

Jon-Corey commented on org.thecrossingchurch.CustomJobs/Jobs/HubspotIntegrationPatching.cs:
```diff
@@ -0,0 +1,442 @@
+            WriteToLog( "Is propsQry Null? : " + propsQry.IsNull().ToString() );
+            WriteToLog( "Is propsQry.results Null? : " + propsQry.results.IsNull().ToString() );
+            WriteToLog( "Is props Null? : " + props.IsNull().ToString() );
+            
```
This line has a bunch of extra whitespace. Can you remove it?

Jon-Corey commented on org.thecrossingchurch.CustomJobs/Jobs/HubspotIntegrationPatching.cs:
```diff
@@ -0,0 +1,442 @@
+            WriteToLog( "Is props Null? : " + props.IsNull().ToString() );
+            
+
+            //Filter to props in Rock Information Group (and Contact information group) 
```
There's some extra white space at the end of this line. Can you remove it?

Jon-Corey commented on org.thecrossingchurch.CustomJobs/Jobs/HubspotIntegrationPatching.cs:
```diff
@@ -0,0 +1,442 @@
+            GetContacts( "https://api.hubapi.com/crm/v3/objects/contacts?limit=100&properties=rock_person_id" );
+            Debug.WriteLine("Contacts returned: " + contacts.Count());
+            WriteToLog(string.Format("Total Contacts to Match: {0}", contacts.Count()));
+            
```
This line has a bunch of extra whitespace. Can you remove it?

Jon-Corey commented on org.thecrossingchurch.CustomJobs/Jobs/HubspotIntegrationPatching.cs:
This line has a bunch of extra whitespace. Can you remove it?

Jon-Corey commented on org.thecrossingchurch.CustomJobs/Jobs/HubspotIntegrationPatching.cs:
This line has a bunch of extra whitespace. Can you remove it?

Jon-Corey commented on org.thecrossingchurch.CustomJobs/Jobs/HubspotIntegrationMatching.cs:
```diff
@@ -0,0 +1,718 @@
+        private int request_count { get; set; }
+        private string businessUnit { get; set; }
+
+        /// <summary> 
```
There's some extra white space at the end of this line. Can you remove it?

Jon-Corey commented on org.thecrossingchurch.CustomJobs/Jobs/HubspotIntegrationPatching.cs:
```diff
@@ -0,0 +1,442 @@
+        }
+
+        /// <summary>
+        /// Job that will run quick SQL queries on a schedule.
```
This XML code comment doesn't seem to be accurate.

Jon-Corey commented on org.thecrossingchurch.CustomJobs/Jobs/HubspotIntegrationMatching.cs:
This line has a bunch of extra whitespace. Can you remove it?

Jon-Corey commented on org.thecrossingchurch.CustomJobs/Jobs/HubspotIntegrationMatching.cs:
This line has a bunch of extra whitespace. Can you remove it?

Jon-Corey commented on org.thecrossingchurch.CustomJobs/Jobs/HubspotIntegrationMatching.cs:
This line has a bunch of extra whitespace. Can you remove it?

Jon-Corey commented on org.thecrossingchurch.CustomJobs/Jobs/HubspotIntegrationMatching.cs:
This line has a bunch of extra whitespace. Can you remove it?

Jon-Corey commented on org.thecrossingchurch.CustomJobs/Jobs/HubspotIntegrationMatching.cs:
There's some extra white space at the end of this line. Can you remove it?

Jon-Corey commented on org.thecrossingchurch.CustomJobs/Jobs/HubspotIntegrationMatching.cs:
This line has a bunch of extra whitespace. Can you remove it?

Jon-Corey commented on org.thecrossingchurch.CustomJobs/Jobs/HubspotIntegrationMatching.cs:
There's some extra white space at the end of this line. Can you remove it?

Jon-Corey commented on org.thecrossingchurch.CustomJobs/Jobs/HubspotIntegrationMatching.cs:
There's some extra white space at the end of this line. Can you remove it?

Jon-Corey commented on org.thecrossingchurch.CustomJobs/Jobs/HubspotIntegrationMatching.cs:
Informational comments like these should have a space after the slashes. Can you update this and similar comments?

shaneLPDev committed:
- Partial Corrections From Code Review

shaneLPDev committed:
- More Corrections From Review

shaneLPDev committed:
- Partial Corrections From Review

shaneLPDev committed:
- Partial Corrections From Review

Jon-Corey commented:
@shaneLPDev, I just reviewed the most recent changes, and they look good. We'll do a test in Train with the latest changes tomorrow and if that goes well, I'll approve this PR and then we'll work on deploying this code.

shaneLPDev committed:
- Corrections From Code Review
